runtime/jvmti/jvmtiThread.c

```diff
@@ -505,16 +511,66 @@ jvmtiGetThreadInfo(jvmtiEnv* env,
 				}
 
 				releaseOMRVMThreadName(targetThread->omrVMThread);
+			}
+
+			if (J9_ARE_ALL_BITS_SET(vm->jclFlags, J9_JCL_FLAG_THREADGROUPS)) {
+				j9object_t group = NULL;
 
-				if (JAVAVM_FROM_ENV(env)->jclFlags & J9_JCL_FLAG_THREADGROUPS) {
 #if JAVA_SPEC_VERSION >= 19
-					j9object_t group = NULL;
-					if (NULL != threadHolder) {
-						group = (j9object_t)J9VMJAVALANGTHREADFIELDHOLDER_GROUP(currentThread, threadHolder);
+				if (isVirtual) {
+					/* For virtual threads, check its state. */
+					if (JVMTI_VTHREAD_STATE_TERMINATED != J9VMJAVALANGVIRTUALTHREAD_STATE(currentThread, threadObject)) {
+						/* The thread group of a virtual thread is always j.l.Thread$Constants.VTHREAD_GROUP. */
+						JNIEnv *jniEnv = (JNIEnv *)currentThread;
+						jclass constants = vm->jlThreadConstants;
+						jfieldID id = vm->vthreadGroupID;
+						jobject tempGroup = NULL;
+
+						vm->internalVMFunctions->internalExitVMToJNI(currentThread);
+						if ((NULL == constants) || (NULL == id)) {
+							/* Cache class and field id. */
+							jclass localRef = (*jniEnv)->FindClass(jniEnv, "java/lang/Thread$Constants");
+							Assert_JVMTI_notNull(localRef);
+							constants = (jclass)((*jniEnv)->NewGlobalRef(jniEnv, localRef));
+							if (NULL == constants) {
+								vm->internalVMFunctions->internalEnterVMFromJNI(currentThread);
```

Comment:
you can add a new label `exit` after release, so you dont have to rearuire vmaccess, just to release immediately

Reply:
Do you mean add label after `internalExitVMToJNI`? Then instead of `vm->internalVMFunctions->internalEnterVMFromJNI(currentThread);` I would have to `releaseVMThread(currentThread, targetThread, thread);`

Reply:
The label will have to be wrapped in #if JAVA_SPEC_VERSION >= 19 as well